### ISSUE TITLE
Use svelte for our js code ?

### ISSUE THREAD

**Issue by mgautierfr** [kiwix/libkiwix#549]
I've recently discovered [Svelte](https://svelte.dev/)
It is a framework to create UI in html/css/js.
In opposition to other framework (vue, react, jquery) svelte works at compilation time and generate output (css/js) only containing what we need and want.

Pro:
- No need to embed libraries (jquery, isotope (#534))
- Only what we need for fast loading
- High level way of programming (component, reactive ui, animation, ...)

Cons:
- Need to generate ("compile") the css/js code along with the cpp code. So we need to setup a node/svelte build env on developer machine (and CI). This could be mitigate by keeping the generated css/js in the git repository. Then, only frontend dev would have to setup de build env. Other dev would simply use the already generated files.

**Comment by kelson42** [comment]
We have discussed this. This is an intetesting proposal. But the main problem we have is our dependence to jquery... and this dependence will vanish anyway.

To me the main question is: which practical improvement will that bring considering our actual Javascript code base?

**Comment by mgautierfr** [comment]
- We would not embed isotope neither.
- As we don't embed a library, we ask user to load (and our server to send) less data. Our binaries don't embed it and so, are smaller.
- As we don't embed a library, we don't have to keep it up to date and we potentially have less security issue

**Comment by kelson42** [comment]
@mgautierfr Ok, so the main advantage is to remove js isotop? If we do so the look&feel woukd be kept? Woukd you recode all the isotop library logic?

**Comment by mgautierfr** [comment]
> so the main advantage is to remove js isotop? 

Yes, but also help us to implement all our frontend (and help us with new feature as incremental search or widget)

> If we do so the look&feel would be kept?

We just have to keep our current css

> Woukd you recode all the isotop library logic?

Mostly no. isotop is just a "flexbox" container with few animations. (and svelte handle animation pretty nicely)

---

Here a small POC of what it is possible to do : https://svelte.dev/repl/2b6fd14cbd454b339dc8c55ea32d67e0?version=3.48.0

Please not that is it is a POC made in less than two hours. Style is not identical. There is no download link and so.
But you can see that:
- It is working (fast)
- Code is pretty clear
- Tiling is made with flexbox, which is standard in css3
- Styling is a copy paste of our `style.css` (+ flexbox)
- No code for transition (but pretty easy to add)
- No history/cookies/... management
- No pagination
- The only files to embed would be the js and css generated

I still have to figure how (and if) we integrate it in our buildsystem (and our cache system for static resources and ...)

**Comment by mgautierfr** [comment]
@juuz0 this could be interesting for you as you work on the frontend also.
Do you have an opinion on svelte ?

**Comment by juuz0** [comment]
I have used svelte little only as opposed to react, jquery so not much opinion :)
But absolutely up for higher level of programming - components. As complexity grows in kiwix-serve, it might reach a point where injecting HTML all the time will be infeasible.
